### PR TITLE
Fix: no-unreachable reporting EmptyStatements (fixes #9081)

### LINKED ISSUE #9081: `no-unreachable` reports EmptyStatements
Thank you for a wonderful tool.  Most times I think I find a bug, instead I learn a JS horror detail.

Can't figure this one out yet.  Don't think it's env related.

* ESLint - 4.3.0
* Node - 8.1.3
* NPM - 5.3.0
* Parser - babel-eslint
* Exhibits on Mac, Win10, Travis CI's debian

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
John-Haugeland:jssm johnhaugeland$ cat .eslintrc 
{

  "parser"  : "babel-eslint",
  "plugins" : [ ],
  "env"     : { "es6": true, "commonjs": true },

  "extends" : "eslint-config-stonecypher"

}
```

`eslint-config-stonecypher` in turn configures every current eslint rule, plus `eslint-plugin-` ... `flowtype`, `fp`, `jsdoc`, `ava`, `unicorn`, `new-with-error`, and `promise`

</details>

**What did you do? Please include the actual source code causing the issue.**

The following code triggers "no-unreachable" at the end of the switch, even though it's the last action in a default, ostensibly because of the `throw`.

```js
const arrow_direction = arrow => {

  switch (arrow) {

    case '->': case '=>': case '~>':
      return 'right';

    case '<-': case '<=': case '<~':
      return 'left';

    case '<->': case '<-=>': case '<-~>':
    case '<=>': case '<=->': case '<=~>':
    case '<~>': case '<~->': case '<~=>':
      return 'both';

    default:
      throw new Error(`arrow_direction: unknown arrow type ${arrow}`);

  };

}
```

**What did you expect to happen?**
Nothing.  No code follows the `throw`.

This similar code does not trigger the warning, I expect because the throw is now outside the switch.

```javascript
const arrow_direction = arrow => {

  switch (arrow) {

    case '->': case '=>': case '~>':
      return 'right';

    case '<-': case '<=': case '<~':
      return 'left';

    case '<->': case '<-=>': case '<-~>':
    case '<=>': case '<=->': case '<=~>':
    case '<~>': case '<~->': case '<~=>':
      return 'both';

  };

  throw new Error(`arrow_direction: unknown arrow type ${arrow}`);

}
```

**What actually happened? Please include the actual, raw output from ESLint.**
```
/Users/johnhaugeland/projects/jssm/src/js/jssm.js
   22:7    warning  'arrow_direction' is assigned a value but never used  no-unused-vars
   22:7    warning  Missing "arrow_direction" variable type annotation    flowtype/require-variable-type
   22:25   warning  Missing return type annotation                        flowtype/require-return-type
   22:25   warning  Missing "arrow" parameter type annotation             flowtype/require-parameter-type
   44:4    warning  Unnecessary semicolon                                 no-extra-semi
   44:4    error    Unreachable code                                      no-unreachable
...
```

It's not like it's a big deal, or anything.  I moved the throws outside of the switch afterwards anyway, because in retrospect it's sorta gross.

It's just surprising, and I wonder whether it's legitimately a subtle bug.

Please have a nice day.

### PR DESCRIPTION
**Pull request by platinumazure**
no-unreachable will only report if there is at least one non-EmptyStatement node in the range

**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

See #9081.

**What changes did you make? (Give an overview)**

no-unreachable now tracks whether non-EmptyStatement nodes have been included in an unreachable code range.

* If an unreachable code range only includes EmptyStatement nodes, the range is not reported. (Rule `no-empty` should be used instead.)
* However, if an unreachable code range has at least one EmptyStatement and at least one non-EmptyStatement, the entire range is reported, including the EmptyStatement(s).

**Is there anything you'd like reviewers to focus on?**

Is there a better way to track empty vs non-empty statements? My preference would have been to iterate on *nodes* (not tokens) between the start and end node of the reporting range and checking that all are EmptyStatements, but SourceCode doesn't have any methods for node iteration. I think what I have works for now, but will fall apart if the traversal order changes.

### CODE REVIEW

**Comment by not-an-aardvark** [comment]
It looks like `no-unreachable` currently has a listener for every statement type. Would it work to simply remove the [listener for `EmptyStatement`](https://github.com/eslint/eslint/blob/a113cd3bf831078c3d7d417aa7ab768b2fbd0fe4/lib/rules/no-unreachable.js#L180)?

**Comment by platinumazure** [comment]
I thought of that, but I'm afraid of it missing some ranges or parts of
ranges that are or contain EmptyStatements. The handler basically does two
things: checks to see if this statement is unreachable and tracks it if so;
and reporting once non-consecutive code is traversed. The first step still
needs to happen for EmptyStatements, I think.

**Comment by platinumazure** [comment]
Adding "do not merge" since there's still some bikeshedding going on.

**Comment by platinumazure** [comment]
@eslint/eslint-team This should be ready for review. Thanks!

@not-an-aardvark Removing the listener will affect the range reported. I personally like the idea of reporting the entire range of unreachable code, as long as we only report code that has more than just EmptyStatements. (Please feel free to look at the unit tests and look at the start/end line/column in the new test cases to see what I'm after.)

**Comment by michaelficarra** [comment]
:-1: I want this rule to continue warning on empty statements. Also, this is a breaking change and thus needs to be behind an option. Using no-empty is not an option for me because I use empty statements which are not dead code, such as in loop bodies.

**Comment by platinumazure** [comment]
I misspoke when I said `no-empty`would help report empty statements, sorry.
The rule for that is `no-extra-semi`. Sorry for the miscommunication and
subsequent confusion.

**Comment by ljharb** [comment]
@michaelficarra What specific code patterns would this rule as-is cover, that wouldn't be covered by no-extra-semi + the-rule-after-this-PR?

**Comment by platinumazure** [comment]
Some examples, in case it helps:

```js
// Case 1: Switch with extra EmptyStatement

switch (foo) {
    case true: throw new Error();
    default: throw new Error();
};         // <-- reported by no-extra-semi and no-unreachable

// After this PR: Line above will not be reported by no-unreachable.
```

```js
// Case 2: Switch with extra EmptyStatement and unreachable non-empty statement

switch (foo) {
    case true: throw new Error();
    default: throw new Error();
};         // <-- reported by no-extra-semi and no-unreachable
bar();     // <-- reported by no-unreachable

// After this PR: No change-- everything from the extra semicolon to the end of `bar();` is reported by no-unreachable
```

The goal of this change is to avoid `no-unreachable` reporting completely frivolous examples (i.e., range contains only EmptyStatement nodes).

**Comment by michaelficarra** [comment]
I'm okay with this change now that I know it's not `no-empty` but `no-extra-semi` which will need to be enabled in its place.

Still, in general, I don't think we should be modifying rule A so that it doesn't report anything that would also be reported by unrelated rule B. **It's fine for one section of code to be identified as wrong in more than one way.** I care more that I can describe what a rule does without a big list of exceptions.

> X: What does the no-unreachable rule do?

> Me: It reports any unreachable code. Unless no-extra-semi would also have reported that code.

> X: Okay, what does no-extra-semi do?

> Me: It reports unnecessary semicolons. Unless ...

(repeat)

**Comment by ljharb** [comment]
The difference here is that empty statements aren't code, in a conceptual sense - people who don't understand parsing and/or spec details aren't going to expect that warning, nor find it valuable.

**Comment by not-an-aardvark** [comment]
> @not-an-aardvark Removing the listener will affect the range reported. I personally like the idea of reporting the entire range of unreachable code, as long as we only report code that has more than just EmptyStatements. (Please feel free to look at the unit tests and look at the start/end line/column in the new test cases to see what I'm after.)

I'm not sure I agree that this is the best behavior. Consider the following code:

```js
switch (foo) {
  case 1: return x;
  default: return y;
};

baz();
```

It seems like the motivation behind this change is that users don't consider `EmptyStatement`s to be "code". So as far as the `no-unreachable` rule is concerned, putting a semicolon after the `switch` statement is a valid stylistic choice. I think it's misleading to make the report range start at the semicolon, because it's only necessary to remove `baz()` to fix the problem, and the rule isn't supposed to care about whether the semicolon is present.

In other words, I think that instead of "reporting the entire range of unreachable code, as long as we only report code that has more than just EmptyStatements", the rule should conceptually just ignore unreachable `EmptyStatement`s entirely. That would provide the clearest error message to the user, and as a bonus it would also make the implementation simpler.

**Comment by platinumazure** [comment]
I'll take another crack at this (just ignoring EmptyStatements entirely) when I get time. Closing this one for now since it's not the desired direction.
